presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
+        superset.getSymbolsWithKnownStatistics().forEach(symbol -> {
+            SymbolStatsEstimate supersetSymbolStats = superset.getSymbolStatistics(symbol);
+            SymbolStatsEstimate leftSymbolStats = left.getSymbolStatistics(symbol);
+            SymbolStatsEstimate rightSymbolStats = right.getSymbolStatistics(symbol);
+            SymbolStatsEstimate intersectSymbolStats = intersect.getSymbolStatistics(symbol);
 
-        double nullsCountLeft = leftStats.getNullsFraction() * leftRowCount;
-        double nullsCountRight = rightStats.getNullsFraction() * rightRowCount;
-        double totalSizeLeft = (leftRowCount - nullsCountLeft) * leftStats.getAverageRowSize();
-        double totalSizeRight = (rightRowCount - nullsCountRight) * rightStats.getAverageRowSize();
-        double newNullsFraction = Math.max(nullsCountLeft - nullsCountRight, 0.0) / newRowCount;
-        double newNonNullsRowCount = newRowCount * (1.0 - newNullsFraction);
+            SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
 
-        StatisticRange range = leftRange;
-        double newDistinctValuesCount = leftStats.getDistinctValuesCount();
-        double leftValuesPerDistinctValue = leftRowCount * (1.0 - leftStats.getNullsFraction()) / leftStats.getDistinctValuesCount();
-        double rightValuesPerDistinctValue = rightRowCount * (1.0 - rightStats.getNullsFraction()) / rightStats.getDistinctValuesCount();
-        if (leftValuesPerDistinctValue <= rightValuesPerDistinctValue) {
-            // right values cover all left values for corresponding distinct values
-            range = strategy.range(leftRange, rightRange);
-            newDistinctValuesCount = leftStats.getDistinctValuesCount() - rightStats.getDistinctValuesCount();
-        }
+            // for simplicity keep the average row size the same as in the input
+            // in most cases the average row size doesn't change after applying filters
+            newSymbolStats.setAverageRowSize(supersetSymbolStats.getAverageRowSize());
 
-        return SymbolStatsEstimate.builder()
-                .setDistinctValuesCount(newDistinctValuesCount)
-                .setHighValue(range.getHigh())
-                .setLowValue(range.getLow())
-                .setAverageRowSize((totalSizeLeft - totalSizeRight) / newNonNullsRowCount)
-                .setNullsFraction(newNullsFraction)
-                .build();
+            // distinctValuesCount, range
+            // assuming distinct values are non-overlapping
+            StatisticRange range = leftSymbolStats.statisticRange().addAndSumDistinctValues(rightSymbolStats.statisticRange());
+            newSymbolStats.setStatisticsRange(range);
+            double supersetDistinctValuesCount = supersetSymbolStats.getDistinctValuesCount();
+            if (range.getDistinctValuesCount() > supersetDistinctValuesCount) {
+                newSymbolStats.setDistinctValuesCount(supersetDistinctValuesCount);
+            }
```

Comment:
there should be check for `newDistinctValueCount==0`.

Reply:
This is responsibility of normalization to set the range to empty, and the nulls fraction to `1` in case of `0` distinct values. We are not duplicating it anywhere else: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java#L74, https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/StatsNormalizer.java#L122